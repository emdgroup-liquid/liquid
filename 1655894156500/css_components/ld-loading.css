:host,svg.ld-loading{--ld-loading-col-base:var(--ld-thm-warning);--ld-loading-col-primary:var(--ld-thm-primary);--ld-loading-col-secondary:var(--ld-thm-secondary);--ld-loading-play-state:running;--ld-loading-size:var(--ld-sp-24);--ld-loading-stretch-dur:4s}:host.ld-loading--neutral,:host:host(.ld-loading--neutral),svg.ld-loading.ld-loading--neutral,svg.ld-loading:host(.ld-loading--neutral){--ld-loading-col-base:var(--ld-col-neutral-100);--ld-loading-col-primary:var(--ld-col-neutral-700);--ld-loading-col-secondary:var(--ld-col-neutral-400)}:host.ld-loading--paused,:host:host(.ld-loading--paused),svg.ld-loading.ld-loading--paused,svg.ld-loading:host(.ld-loading--paused){--ld-loading-play-state:paused}:host,svg.ld-loading{fill:none;border-radius:100%;display:inline-flex;flex-shrink:0;height:var(--ld-loading-size);-webkit-mask-image:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 100 100"><defs/><circle cx="50" cy="50" r="40" stroke="currentColor" stroke-width="18"/></svg>');mask-image:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 100 100"><defs/><circle cx="50" cy="50" r="40" stroke="currentColor" stroke-width="18"/></svg>');overflow:hidden;width:var(--ld-loading-size)}:host circle,svg.ld-loading circle{stroke:var(--ld-loading-col-base);stroke-width:40;stroke-dashoffset:330;stroke-linecap:round;transform-origin:center}:host g circle,svg.ld-loading g circle{stroke-dasharray:570;animation:ld-loading-stretch var(--ld-loading-stretch-dur) ease infinite var(--ld-loading-play-state)}:host g circle:last-of-type,svg.ld-loading g circle:last-of-type{--ld-stroke-dashoffset-from:420;--ld-stroke-dashoffset-to:540}:host g circle:nth-last-of-type(2),svg.ld-loading g circle:nth-last-of-type(2){--ld-stroke-dashoffset-from:300;--ld-stroke-dashoffset-to:490;stroke:var(--ld-loading-col-secondary);stroke-width:39;animation-delay:calc(var(--ld-loading-stretch-dur)*.2)}:host g circle:nth-last-of-type(3),svg.ld-loading g circle:nth-last-of-type(3){--ld-stroke-dashoffset-from:330;--ld-stroke-dashoffset-to:450;stroke:var(--ld-loading-col-primary);stroke-width:39;animation-delay:calc(var(--ld-loading-stretch-dur)*.4)}@keyframes ld-loading-stretch{0%{stroke-dashoffset:var(--ld-stroke-dashoffset-from)}50%{stroke-dashoffset:var(--ld-stroke-dashoffset-to)}to{stroke-dashoffset:var(--ld-stroke-dashoffset-from)}}