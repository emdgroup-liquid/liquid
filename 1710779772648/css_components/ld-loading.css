:host,svg.ld-loading{--ld-loading-col-base:var(--ld-thm-warning);--ld-loading-col-primary:var(--ld-thm-primary);--ld-loading-col-secondary:var(--ld-thm-secondary);--ld-loading-play-state:running;--ld-loading-size:var(--ld-sp-24);--ld-loading-stretch-dur:4s}:host.ld-loading--neutral,:host:host(.ld-loading--neutral),svg.ld-loading.ld-loading--neutral,svg.ld-loading:host(.ld-loading--neutral){--ld-loading-col-base:var(--ld-col-neutral-100);--ld-loading-col-primary:var(--ld-col-neutral-700);--ld-loading-col-secondary:var(--ld-col-neutral-400)}:host.ld-loading--paused,:host:host(.ld-loading--paused),svg.ld-loading.ld-loading--paused,svg.ld-loading:host(.ld-loading--paused){--ld-loading-play-state:paused}:host,svg.ld-loading{display:inline-flex;flex-shrink:0;height:var(--ld-loading-size);width:var(--ld-loading-size);fill:none;border-radius:100%;-webkit-mask-image:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 100 100"><defs/><circle cx="50" cy="50" r="40" stroke="currentColor" stroke-width="18"/></svg>');mask-image:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 100 100"><defs/><circle cx="50" cy="50" r="40" stroke="currentColor" stroke-width="18"/></svg>');overflow:hidden}:host circle,svg.ld-loading circle{transform-origin:center;stroke:var(--ld-loading-col-base);stroke-width:40;stroke-dashoffset:330;stroke-linecap:round}:host g circle,svg.ld-loading g circle{stroke-dasharray:570;animation:ld-loading-stretch var(--ld-loading-stretch-dur) ease infinite var(--ld-loading-play-state)}:host g circle:last-of-type,svg.ld-loading g circle:last-of-type{--ld-stroke-dashoffset-from:420;--ld-stroke-dashoffset-to:540}:host g circle:nth-last-of-type(2),svg.ld-loading g circle:nth-last-of-type(2){--ld-stroke-dashoffset-from:300;--ld-stroke-dashoffset-to:490;stroke:var(--ld-loading-col-secondary);animation-delay:calc(.2 * var(--ld-loading-stretch-dur));stroke-width:39}:host g circle:nth-last-of-type(3),svg.ld-loading g circle:nth-last-of-type(3){--ld-stroke-dashoffset-from:330;--ld-stroke-dashoffset-to:450;stroke:var(--ld-loading-col-primary);animation-delay:calc(.4 * var(--ld-loading-stretch-dur));stroke-width:39}@keyframes ld-loading-stretch{0%{stroke-dashoffset:var(--ld-stroke-dashoffset-from)}50%{stroke-dashoffset:var(--ld-stroke-dashoffset-to)}to{stroke-dashoffset:var(--ld-stroke-dashoffset-from)}}