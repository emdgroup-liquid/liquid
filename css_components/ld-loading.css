.ld-loading,:host{--ld-loading-size:1.5rem;--ld-loading-col:var(--ld-col-neutral-300);background:var(--ld-loading-col);height:var(--ld-loading-size);width:var(--ld-loading-size);display:inline-flex;flex-shrink:0;-webkit-mask-image:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 23 23"><defs/><circle cx="11.5" cy="11.5" r="9.5" stroke="currentColor" stroke-dasharray="2.55 5" stroke-width="4"/></svg>');mask-image:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 23 23"><defs/><circle cx="11.5" cy="11.5" r="9.5" stroke="currentColor" stroke-dasharray="2.55 5" stroke-width="4"/></svg>');animation:ld-loading-animation 1.5s linear infinite}@keyframes ld-loading-animation{0%{transform:rotate(0deg)}to{transform:rotate(1turn)}}