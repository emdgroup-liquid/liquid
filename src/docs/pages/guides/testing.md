---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

