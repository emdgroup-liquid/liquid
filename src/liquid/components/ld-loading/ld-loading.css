/**
 * @prop --ld-loading-size: Size of the loading indicator (1.5rem by default).
 */
:host,
.ld-loading {
  --ld-loading-size: 1.5rem;
  height: var(--ld-loading-size);
  width: var(--ld-loading-size);
  display: inline-flex;
  flex-shrink: 0;
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGZpbGw9Im5vbmUiIHZpZXdCb3g9IjAgMCAyMyAyMyI+PGRlZnMvPjxjaXJjbGUgY3g9IjExLjUiIGN5PSIxMS41IiByPSI5LjUiIHN0cm9rZT0iI0JGQkZCRiIgc3Ryb2tlLWRhc2hhcnJheT0iMi41NSA1IiBzdHJva2Utd2lkdGg9IjQiLz48L3N2Zz4=');
  background-size: 100%;
  background-repeat: no-repeat;
  animation: ld-loading-animation 1.5s linear infinite;
}

@keyframes ld-loading-animation {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
