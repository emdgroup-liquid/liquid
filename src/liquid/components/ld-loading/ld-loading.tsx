import { Component, h, Host, Prop } from "@stencil/core";
import { getClassNames } from "../../utils/getClassNames";

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: "ld-loading",
  styleUrl: "ld-loading.css",
  shadow: true,
})
export class LdLoading {
  /** Used as svg title element content. */
  @Prop() label? = "Loading";

  /** Uses neutral colors. */
  @Prop() neutral?: boolean;

  /** Pauses all animations. */
  @Prop() paused?: boolean;

  render() {
    const cl = getClassNames([
      "ld-loading",
      this.neutral && "ld-loading--neutral",
      this.paused && "ld-loading--paused",
    ]);

    return (
      <Host class={cl}>
        <svg viewBox="0 0 100 100" preserveAspectRatio="xMidYMid meet">
          <title>{this.label}</title>
          <circle cx="50" cy="50" r="50" />
          <g>
            <circle cx="50" cy="50" r="50" />
            <circle cx="50" cy="50" r="50" />
            <circle cx="50" cy="50" r="50" />
            {!this.paused && (
              // When zooming in safari CSS transforms get messed up.
              // That is why we need to use an SVG animation here.
              <animateTransform
                attributeName="transform"
                attributeType="XML"
                type="rotate"
                from="0 50 50"
                to="360 50 50"
                dur="0.9s"
                repeatCount="indefinite"
              />
            )}
          </g>
        </svg>
      </Host>
    );
  }
}
