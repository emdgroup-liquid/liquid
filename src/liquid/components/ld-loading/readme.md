---
eleventyNavigation:
  key: Loading Indicator
  parent: Components
layout: layout.njk
title: Loading Indicator
permalink: components/ld-loading/
---

<link rel="stylesheet" href="css_components/ld-loading.css">

# ld-loading

Use the `ld-loading` component to indicate that the user should wait for a process to complete. Combine it with textual information on what is going on for a better user experience. 

## Basic

{% example %}
<ld-loading></ld-loading>

<!-- CSS component -->

<svg class="ld-loading" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid meet">
  <title>Loading</title>
  <circle cx="50" cy="50" r="50" />
  <g>
    <circle cx="50" cy="50" r="50" />
    <circle cx="50" cy="50" r="50" />
    <circle cx="50" cy="50" r="50" />
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      type="rotate"
      from="0 50 50"
      to="360 50 50"
      dur="0.9s"
      repeatCount="indefinite"
    />
  </g>
</svg>
{% endexample %}

## Neutral

{% example %}
<ld-loading neutral></ld-loading>

<!-- CSS component -->

<svg class="ld-loading ld-loading--neutral" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid meet">
  <title>Loading</title>
  <circle cx="50" cy="50" r="50" />
  <g>
    <circle cx="50" cy="50" r="50" />
    <circle cx="50" cy="50" r="50" />
    <circle cx="50" cy="50" r="50" />
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      type="rotate"
      from="0 50 50"
      to="360 50 50"
      dur="0.9s"
      repeatCount="indefinite"
    />
  </g>
</svg>
{% endexample %}

## CSS Variables

| Variable                     | Description                               | Default                   |
| ---------------------------- | ----------------------------------------- | ------------------------- |
| `--ld-loading-col-base`      | Base color of the loading indicator.      | `var(--ld-thm-warning)`   |
| `--ld-loading-col-primary`   | Primary color of the loading indicator.   | `var(--ld-thm-primary)`   |
| `--ld-loading-col-secondary` | Secondary color of the loading indicator. | `var(--ld-thm-secondary)` |
| `--ld-loading-size`          | Size of the loading indicator.            | `var(--ld-sp-24)`         |

<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description                                              | Type               | Default     |
| --------- | --------- | -------------------------------------------------------- | ------------------ | ----------- |
| `key`     | `key`     | for tracking the node's identity when working with lists | `string \| number` | `undefined` |
| `label`   | `label`   | Used as svg title element content.                       | `string`           | `'Loading'` |
| `neutral` | `neutral` | Uses neutral colors.                                     | `boolean`          | `undefined` |
| `paused`  | `paused`  | Pauses all animations.                                   | `boolean`          | `undefined` |
| `ref`     | `ref`     | reference to component                                   | `any`              | `undefined` |


----------------------------------------------

 
