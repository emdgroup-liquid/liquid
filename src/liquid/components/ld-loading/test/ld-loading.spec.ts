import { newSpecPage } from '@stencil/core/testing'
import { LdLoading } from '../ld-loading'

describe('ld-loading', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdLoading],
      html: `<ld-loading />`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
