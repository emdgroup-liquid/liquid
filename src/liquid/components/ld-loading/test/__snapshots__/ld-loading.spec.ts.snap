// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-loading is neutral 1`] = `
<ld-loading class="ld-loading ld-loading--neutral" neutral="">
  <mock:shadow-root>
    <svg preserveAspectRatio="xMidYMid meet" viewBox="0 0 100 100">
      <title>
        Loading
      </title>
      <circle cx="50" cy="50" r="50"></circle>
      <g>
        <circle cx="50" cy="50" r="50"></circle>
        <circle cx="50" cy="50" r="50"></circle>
        <circle cx="50" cy="50" r="50"></circle>
        <animateTransform attributeName="transform" attributeType="XML" dur="0.9s" from="0 50 50" repeatCount="indefinite" to="360 50 50" type="rotate"></animateTransform>
      </g>
    </svg>
  </mock:shadow-root>
</ld-loading>
`;

exports[`ld-loading renders 1`] = `
<ld-loading class="ld-loading">
  <mock:shadow-root>
    <svg preserveAspectRatio="xMidYMid meet" viewBox="0 0 100 100">
      <title>
        Loading
      </title>
      <circle cx="50" cy="50" r="50"></circle>
      <g>
        <circle cx="50" cy="50" r="50"></circle>
        <circle cx="50" cy="50" r="50"></circle>
        <circle cx="50" cy="50" r="50"></circle>
        <animateTransform attributeName="transform" attributeType="XML" dur="0.9s" from="0 50 50" repeatCount="indefinite" to="360 50 50" type="rotate"></animateTransform>
      </g>
    </svg>
  </mock:shadow-root>
</ld-loading>
`;

exports[`ld-loading uses custom label 1`] = `
<ld-loading class="ld-loading ld-loading--neutral" label="Doing stuff..." neutral="">
  <mock:shadow-root>
    <svg preserveAspectRatio="xMidYMid meet" viewBox="0 0 100 100">
      <title>
        Doing stuff...
      </title>
      <circle cx="50" cy="50" r="50"></circle>
      <g>
        <circle cx="50" cy="50" r="50"></circle>
        <circle cx="50" cy="50" r="50"></circle>
        <circle cx="50" cy="50" r="50"></circle>
        <animateTransform attributeName="transform" attributeType="XML" dur="0.9s" from="0 50 50" repeatCount="indefinite" to="360 50 50" type="rotate"></animateTransform>
      </g>
    </svg>
  </mock:shadow-root>
</ld-loading>
`;
